### ISSUE TITLE
how to use nvrtc to run a sm90 kernel

### ISSUE THREAD

**Issue by mengchihe** [NVIDIA/cutlass#1889]
Hi

I want to use nvrtc to compile a sm90 kernel in runtime. The problem is that I don't have the kernel instance on host thus can't run to_underlying_arguments to get kernel param to launch the kernel.

When running a sm80 code, I can get it's params directly since Params is equal to Arguments, but for a sm90 kernel I needs to construct tma_load struct before I launch the kernel.

I see there is a nvrtc example in cutlass which use python to launch a nvcc progress to compile host api for the kernel such as initialize into a .so. But I have a c++ application and can't launch any child progress in it. 

So is there anyway to runtime compile a sm90 kernel in my application. I see make_tma_copy is defined as CUDA_HOST_RTC which can be used in nvrtc, but it will call cuTensorMapEncodeTiled which can only called on host, and nvrtcCompileProgram can only compile a kernel function. Is there any example to use make_tma_copy in nvrtc, thanks.

**Comment by mengchihe** [comment]
I found out that make_tma_copy will not construct tma_desc when use nvrtc, is the right way to run sm90 kernel in rtc is construct tma_desc on host, call to_underlying_argument in kernel and then copy tma_desc to tma_load in Params?

**Comment by thakkarV** [comment]
@jackkosaian has tried this before and may be able to help here

**Comment by ghostplant** [comment]
I have the similar question. For the current interface design, `tma = make_tma_atom(make_tensor(dptr_A), ..)` is designed to calculate on the host source code and then its result will be passed to `kernelLaunch<<<>>>(.. , &tma)` as an argument whose content is opaque.

Is it possible to construct kernel-required `tma` within kernel source behind NVRTC, so that host code can normally launch parameters without recomputing tma data-structures, e.g. `kernelLaunch<<<>>>(.. , &dptr_A)`

**Comment by jackkosaian** [comment]
> Is it possible to construct kernel-required tma within kernel source behind NVRTC, so that host code can normally launch parameters without recomputing tma data-structures, e.g. kernelLaunch<<<>>>(.. , &dptr_A)

This is not currently possible.

We do not have a technique at the moment for constructing TMA descriptors outside of the host-based method.

**Comment by jackkosaian** [comment]
I misspoke in the comment above. It is possible to construct Tensor Maps on device. [Here ](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html?highlight=Encoding%2520a%2520tensor%2520map%2520on%2520device#encoding-a-tensor-map-on-device) is some documentation on this.

However, this has not been experimented with in CUTLASS.

**Comment by mengchihe** [comment]
> I misspoke in the comment above. It is possible to construct Tensor Maps on device. [Here ](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html?highlight=Encoding%2520a%2520tensor%2520map%2520on%2520device#encoding-a-tensor-map-on-device) is some documentation on this.
> 
> However, this has not been experimented with in CUTLASS.

Thanks for your reply.

This method seems similar to what cutlass does in group gemm, construct a basic tensor map on host, transfer to kernel in params, adjust its attributes such as address in kernel to get a new tensor map, than use this new tensor map. Seems still needs construct a basic tensor map on host.

sm90 kernel in cutlass do tensor map construction in to_underlying_arguments which will be called on host, but I don't have the kernel instance when use nvrtc thus can't call this function. Seems the only way is to construct a basic or actual tensor map on host and put it into kernel Params, call to_underlying_arguments in kernel which will do all the thing to convert Arguments to Params except tensor map construction, and then copy input tensor map into tma_load_a/b/c/d.

Another problem is how to get grid dimension and shared memory size when use rtc, I write some code to get this still because I don't have the kernel instance, is there anyway that I can reuse code in cutlass to get this. Thanks

**Comment by ghostplant** [comment]
> > I misspoke in the comment above. It is possible to construct Tensor Maps on device. [Here ](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html?highlight=Encoding%2520a%2520tensor%2520map%2520on%2520device#encoding-a-tensor-map-on-device) is some documentation on this.
> > However, this has not been experimented with in CUTLASS.
> 
> Thanks for your reply.
> 
> This method seems similar to what cutlass does in group gemm, construct a basic tensor map on host, transfer to kernel in params, adjust its attributes such as address in kernel to get a new tensor map, than use this new tensor map. Seems still needs construct a basic tensor map on host.
> 
> sm90 kernel in cutlass do tensor map construction in to_underlying_arguments which will be called on host, but I don't have the kernel instance when use nvrtc thus can't call this function. Seems the only way is to construct a basic or actual tensor map on host and put it into kernel Params, call to_underlying_arguments in kernel which will do all the thing to convert Arguments to Params except tensor map construction, and then copy input tensor map into tma_load_a/b/c/d.
> 
> Another problem is how to get grid dimension and shared memory size when use rtc, I write some code to get this still because I don't have the kernel instance, is there anyway that I can reuse code in cutlass to get this. Thanks

I don't think this is a problem. You just need to get the `ABI name` of your kernel function and use cuModuleGetFunction() to get the function handler, then based on the function handler you can further set max shared memory and cluster launch config.

**Comment by ghostplant** [comment]
> I misspoke in the comment above. It is possible to construct Tensor Maps on device. [Here ](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html?highlight=Encoding%2520a%2520tensor%2520map%2520on%2520device#encoding-a-tensor-map-on-device) is some documentation on this.
> 
> However, this has not been experimented with in CUTLASS.

I'm not sure if those documents are up-to-date. Here is my code:

```
using TmaA = cute::Copy_Atom<cute::Copy_Traits<cute::SM90_TMA_LOAD, cute::C<131072>, cute::AuxTmaParams<cute::tuple<cute::ScaledBasis<cute::C<1>, 1>, cute::ScaledBasis<cute::C<1>, 0> >, cute::Layout<cute::tuple<cute::C<64>, cute::C<128> >, cute::tuple<cute::ScaledBasis<cute::C<1>, 1>, cute::ScaledBasis<cute::C<1>, 0> > > const&, cute::Swizzle<3, 4, 3> const&> >, cutlass::half_t>;

printf("%ld\n", sizeof(TmaA));  // I get 192-Byte structure passing to kernel as an argument, while the link your provided above indicates sizeof(struct tensormap_params) == 124.
```

Do you think sizeof(TmaA) == 192 on my case is reasonable? How to consult its member's storage order correctly?

**Comment by mengchihe** [comment]
> > > I misspoke in the comment above. It is possible to construct Tensor Maps on device. [Here ](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html?highlight=Encoding%2520a%2520tensor%2520map%2520on%2520device#encoding-a-tensor-map-on-device) is some documentation on this.
> > > However, this has not been experimented with in CUTLASS.
> > 
> > 
> > Thanks for your reply.
> > This method seems similar to what cutlass does in group gemm, construct a basic tensor map on host, transfer to kernel in params, adjust its attributes such as address in kernel to get a new tensor map, than use this new tensor map. Seems still needs construct a basic tensor map on host.
> > sm90 kernel in cutlass do tensor map construction in to_underlying_arguments which will be called on host, but I don't have the kernel instance when use nvrtc thus can't call this function. Seems the only way is to construct a basic or actual tensor map on host and put it into kernel Params, call to_underlying_arguments in kernel which will do all the thing to convert Arguments to Params except tensor map construction, and then copy input tensor map into tma_load_a/b/c/d.
> > Another problem is how to get grid dimension and shared memory size when use rtc, I write some code to get this still because I don't have the kernel instance, is there anyway that I can reuse code in cutlass to get this. Thanks
> 
> I don't think this is a problem. You just need to get the `ABI name` of your kernel function and use cuModuleGetFunction() to get the function handler, then based on the function handler you can further set max shared memory and cluster launch config.

I mean how to get shared memory and grid size, not how to set kernel attribute
